### PR TITLE
Silent Wings SITL for autonomous soaring

### PR DESCRIPTION
**Pull request by akolobov**
This change allows ArduPlane to run a SITL with Silent Wings, one of the best commercially available soaring flight simulators. Missions and parameter files for some of Silent Wings' aircraft are in the libraries/SITL/examples/SilentWings directory. Instructions for running the Silent Wings SITL will be included with the rest of ArduPlane documentation. 

This functionality comes from, and has been tested as part of https://github.com/Microsoft/Frigatebird using both demo and full versions of Silent Wings. Glitches I've noticed when testing it as part of ArduPlane:

- If the aircraft has no mission is loaded and the aircraft is asked to enter AUTO, the SITL crashes.

- Setting SOAR_ENABLE to 1 sometimes crashes SITL as well. (Setting this parameter typically needs to be done only once per airframe though.) 

Thanks to @samuelctabor for his help with this addition.

### CODE REVIEW

File: libraries/AP_InertialSensor/AP_InertialSensor.cpp [hunk 1]
```diff
@@ -699,6 +699,11 @@ AP_InertialSensor::detect_backends(void)
         ADD_BACKEND(AP_InertialSensor_HIL::detect(*this));
         return;
     }
+
+    if (_sitl_mode) {
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor.cpp]: why is this needed?

File: libraries/AP_InertialSensor/AP_InertialSensor.cpp [hunk 2]
```diff
@@ -1039,7 +1044,7 @@ bool AP_InertialSensor::calibrate_trim(float &trim_roll, float &trim_pitch)
 bool AP_InertialSensor::accel_calibrated_ok_all() const
 {
     // calibration is not applicable for HIL mode
-    if (_hil_mode) {
+    if (_hil_mode || _sitl_mode) {
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor.cpp]: this would break using SITL for testing accelcal

File: libraries/SITL/SIM_SilentWings.cpp [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+    first_pkt_timestamp_ms(0),
+    time_base_us(0),
+    sock(true),
+    home_initialized(false),
+    inited_first_pkt_timestamp(false)
+{
+    // Try to bind to a specific port so that if we restart ArduPilot
+    // SilentWings keeps sending us packets. Not strictly necessary but
+    // useful for debugging
+    sock.bind("127.0.0.1", 6060);
```
Comment (tridge) [inline libraries/SITL/SIM_SilentWings.cpp]: this means you can't run silentwings and ardupilot on different computers

**Comment by tridge** [comment]
I've just bought SilentWings to test it. Unfortunately it segfaults on my ubuntu 18.04 linux box. I've sent a bug report to the support address

File: libraries/AP_InertialSensor/AP_InertialSensor.cpp (hunk 2 above)
Comment (akolobov) [reply libraries/AP_InertialSensor/AP_InertialSensor.cpp]: Got it... Well, AP_InertialSensor is unaware of EKF type, so the only other way of having AP_InertialSensor report OK health with 0 offsets in SITL is to pass EKF type to accel_calibrated_ok_all(). Are you OK with that solution?

File: libraries/SITL/SIM_SilentWings.cpp (hunk 3 above)
Comment (akolobov) [reply libraries/SITL/SIM_SilentWings.cpp]: Right, will fix in the upcoming iteration.

File: libraries/AP_InertialSensor/AP_InertialSensor.cpp (hunk 2 above)
Comment (tridge) [reply libraries/AP_InertialSensor/AP_InertialSensor.cpp]: it could have an #ifdef for SITL and check EKF type. Similar to what I did for this PR:
https://github.com/ArduPilot/ardupilot/pull/11634

File: libraries/AP_InertialSensor/AP_InertialSensor.cpp (hunk 1 above)
Comment (akolobov) [reply libraries/AP_InertialSensor/AP_InertialSensor.cpp]: This is to prevent AP_InertialSensor from reporting poor health when in SITL with 0 accelerometer offsets. Per your comment below, looks like this approach throws out the baby with the water though...

**Comment by tridge** [comment]
I now I SilentWings on Linux not crashing, but I don't see how to start the networking. I can fly it fine with a joystick, but it hasn't opened any network ports. The code is clearly expecting for it to start listening on UDP 6070. How do I enable that?

**Comment by samuelctabor** [comment]
Hi @tridge 

You need to enable it in the configuration file options.dat; see http://wiki.silentwings.no/index.php/UDP_Output and 
http://wiki.silentwings.no/index.php/Remote_Control

It's been a while but I remember I had to actually disconnect the joystick, as it interfered with the SITL sent controls.

How did you get your installation working, as I also have the segfault issue since upgrading. I finally got a reply from support suggesting deleting some library files - did that do it for you?

**Comment by peterbarker** [comment]
On Sun, 23 Jun 2019, Andrey Kolobov wrote:

>     If the aircraft has no mission is loaded and the aircraft is asked to
>     enter AUTO, the SITL crashes.

That's a bit of a worry.  Is it reproducible on master with normal SITL? 
I'm unable to reproduce here.

  Which version of ArduPlane?  Can you provide steps to reproduce, a stack 
trace, or perhaps some more information, please?

>     Setting SOAR_ENABLE to 1 sometimes crashes SITL as well. (Setting this
>     parameter typically needs to be done only once per airframe though.)

Also a rather unacceptable circumstance.  Are you able to run under 
valgrind and/or gdb to provode more information here?  I'm also unable to 
reproduce any issue locally here.

Thanks,

Peter

**Comment by tridge** [comment]
> You need to enable it in the configuration file options.dat; see http://wiki.silentwings.no/index.php/UDP_Output and

ok, we should add a README.md in libraries/SITL/examples/SilentWings explaining how to get it setup. We'll also need a wiki page

> How did you get your installation working, as I also have the segfault issue since upgrading. I finally got a reply from support suggesting deleting some library files - did that do it for you?

yes, that worked for me

File: libraries/AP_InertialSensor/AP_InertialSensor.h [hunk 4]
```diff
@@ -504,6 +507,9 @@ class AP_InertialSensor : AP_AccelCal_Client
     // are we in HIL mode?
     bool _hil_mode:1;
 
+    // are we in SITL mode?
+    bool _sitl_mode:1;
```
Comment (magicrub) [inline libraries/AP_InertialSensor/AP_InertialSensor.h]: like what @tridge said, better to use #if CONFIG_HAL_BOARD == HAL_BOARD_SITL

File: libraries/SITL/SIM_SilentWings.cpp [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+    char *buf = nullptr;
+    // Turn off direct joystick input to the simulator. All joystick commands
+    // should go through Mission Planner and get properly fused with ArduPlane's
+    // control inputs when in automatic flight modes.
+    float joystick = 0.0f;
+    float aileron  = filtered_servo_angle(input, 0);
+    float elevator = filtered_servo_angle(input, 1);
+    float throttle = filtered_servo_range(input, 2);
+    float rudder   = filtered_servo_angle(input, 3);
+    float flap     = 0.0f;
```
Comment (magicrub) [inline libraries/SITL/SIM_SilentWings.cpp]: couldn't this check the current flap state?

Comment (magicrub) [reply libraries/SITL/SIM_SilentWings.cpp]: good change for next upcoming iteration instead of this initial one. Best to get the big chunks in a known-working state without scope creep

File: libraries/SITL/SIM_SilentWings.cpp [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+  This is a blocking function
+ */
+bool SilentWings::recv_fdm(void)
+{
+    fdm_packet tmp_pkt;
+    memset(&pkt, 0, sizeof(pkt));
+
+    ssize_t nread = sock.recv(&tmp_pkt, sizeof(pkt), 0);
+    
+    // nread == -1 (255) means no data has arrived
```
Comment (magicrub) [inline libraries/SITL/SIM_SilentWings.cpp]: comment says it's 255 which is wrong. ssize_t is signed size_t so it's actually -1.

File: libraries/SITL/SIM_SilentWings.cpp [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+    }
+    else {
+        if (use_time_sync) {
+            sync_frame_time();
+        }
+    }
+    
+    update_mag_field_bf();
+    
+    int32_t now = AP_HAL::millis();
```
Comment (magicrub) [inline libraries/SITL/SIM_SilentWings.cpp]: needs to be uint32_t

File: libraries/SITL/SIM_SilentWings.h [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+           float  vario;                    // m/s       Total energy-compensated variometer
+           float  heading;                  // Degrees   Compass heading
+           float  rate_of_turn;             // Deg/sec   Rate of turn
+           float  airpressure;              // Pa        Local air pressure at aircraft altitude)
+           float  density;                  // Air density at aircraft altitude
+           float  temperature;              // Degrees Celcius   Air temperature at aircraft altitude
+    } pkt;
+    
+    struct {
+        int32_t last_report_ms;
```
Comment (magicrub) [inline libraries/SITL/SIM_SilentWings.h]: uint32_t

**Comment by akolobov** [comment]
@tridge, detailed instructions for running Silent Wings SITL are here: https://github.com/microsoft/Frigatebird/blob/master/SITL.md. They are for Windows, but the approach should be the same; as @samuelctabor mentioned, you need to edit options.dat (specifically, its cached copy).

I was actually planning to check a version of those instructions into ArduPilot documentation, so comments on them would be appreciated.

File: libraries/AP_InertialSensor/AP_InertialSensor.cpp (hunk 2 above)
Comment (akolobov) [reply libraries/AP_InertialSensor/AP_InertialSensor.cpp]: OK, looks like that PR fixed this, I'll revert my change to AP_InertialSensor.cpp

File: libraries/AP_InertialSensor/AP_InertialSensor.h (hunk 4 above)
Comment (akolobov) [reply libraries/AP_InertialSensor/AP_InertialSensor.h]: Will revert out the entire change to AP_InertialSensor.

File: libraries/SITL/SIM_SilentWings.cpp (hunk 6 above)
Comment (akolobov) [reply libraries/SITL/SIM_SilentWings.cpp]: Yes, but if someone decides to print this value an unsigned, they'll see 255; that's what the comment is referring to.

File: libraries/SITL/SIM_SilentWings.cpp (hunk 5 above)
Comment (akolobov) [reply libraries/SITL/SIM_SilentWings.cpp]: > couldn't this check the current flap state?

Sure, will fix in the PR update.

> good change for next upcoming iteration instead of this initial one. Best to get the big chunks in a known-working state without scope creep

Sorry, not sure what this is referring to. This is the known working state -- e.g., the SITL for JSBSim works in the same way.

File: libraries/SITL/SIM_SilentWings.h (hunk 8 above)
Comment (akolobov) [reply libraries/SITL/SIM_SilentWings.h]: Fixed

File: libraries/SITL/SIM_SilentWings.cpp (hunk 7 above)
Comment (akolobov) [reply libraries/SITL/SIM_SilentWings.cpp]: Fixed

**Comment by akolobov** [comment]
@peterbarker I've reverted out some of the changes and can't repro the aforementioned crashes anymore. They also don't happen in the original version of this SITL in https://github.com/Microsoft/Frigatebird

**Comment by samuelctabor** [comment]
Tested and working pretty well for me :)

Some problem with getting nice smooth flying but I don't believe they are related to the simulator interface.

**Comment by tridge** [comment]
I've rebased this and squashed the commits. This can be merged when CI passes

**Comment by tridge** [comment]
now we need a wiki page, similar to what we have for flightaxis here:
http://ardupilot.org/dev/docs/sitl-with-realflight.html
